DiffPlex/IDiffer.cs

```diff
@@ -1,20 +1,20 @@
-﻿using System;
-using DiffPlex.Model;
+﻿using DiffPlex.Model;
 
 namespace DiffPlex
 {
     /// <summary>
-    /// Provides methods for generate differences between texts
+    /// Responsible for generating differences between texts
     /// </summary>
     public interface IDiffer
     {
-        DiffResult CreateLineDiffs(string oldText, string newText, bool ignoreWhitespace);
-        DiffResult CreateLineDiffs(string oldText, string newText, bool ignoreWhitespace, bool ignoreCase);
-        DiffResult CreateCharacterDiffs(string oldText, string newText, bool ignoreWhitespace);
-        DiffResult CreateCharacterDiffs(string oldText, string newText, bool ignoreWhitespace, bool ignoreCase);
-        DiffResult CreateWordDiffs(string oldText, string newText, bool ignoreWhitespace, char[] separators);
-        DiffResult CreateWordDiffs(string oldText, string newText, bool ignoreWhitespace, bool ignoreCase, char[] separators);
-        DiffResult CreateCustomDiffs(string oldText, string newText, bool ignoreWhiteSpace, Func<string, string[]> chunker);
-        DiffResult CreateCustomDiffs(string oldText, string newText, bool ignoreWhiteSpace, bool ignoreCase, Func<string, string[]> chunker);
+        /// <summary>
+        /// Create a diff by comparing text line by line
+        /// </summary>
+        /// <param name="oldText">The old text.</param>
+        /// <param name="newText">The new text.</param>
+        /// <param name="ignoreWhiteSpace">if set to <c>true</c> will ignore white space when determining if lines are the same.</param>
+        /// <param name="ignoreCase">Determine if the text comparision is case sensitive or not</param>
+        /// <returns>A DiffResult object which details the differences</returns>
+        DiffResult CreateDiffs(string oldText, string newText, bool ignoreWhiteSpace, bool ignoreCase);
```

Comment:
I agreed that this is a cleaner interface but I think we should avoid a breaking change here if possible. This library has been extremely stable in its api and unless there is a truly compelling reason to break it I think we should keep it consistent. We can talk more about what the right approach is to make your good changes in a non-breaking way

Reply:
I would like to avoid breaking changes, and I hope that this could be achieved with extension methods. However, after the final refactoring, all those methods from the previous `IDiffer` interface do not have sense anymore. I think you should bump up the major number of the library version - this is why somebody came up with the idea of the major number, to communicate breaking changes.  I think users of this library should without the problems migrate to the new major version - I didn't make any changes to the functionality, I change only how the code is organized.

Reply:
We could instead have IChunker be a function argument to a new method in IDiffer. The other existing methods can make use of the new chunker implementations internally and the new method will be the generic version.

Reply:
**Option 1**: Add a new method to `IDiffer` that accepts `IChunker` and convert the existing methods into extension methods - thanks to that we can clean up `IDiffer` interface and if somebody will want to provide a new implementation then he will not be obligated to implement all the overloads.  But this approach enforces to add extra methods to `ISideBySideDiffBuilder` and `IInlineDiffBuilder` to allow for providing custom `IChunker` (we also need to hardcode a specific implementation of `IChunker` for existing methods).  This is how things work in the procedural language - instead of encapsulating `IChunker` inside the `IDiffer` we need to add an extra dependency to all `IDiffer` consumers. I think this makes the overall design more clutter than leads to simplification.

**Option 2**: Just bump up the major version. I don't know why people afraid so much of introducing breaking changes especially when it leads to better design and extensibility. Avoiding it slows down the progress and cause code quality degradation. Right now I feel that we are forced to hack in order to preserve API backward compatibility. 


You are the owner - make a call.

Reply:
Doing the extension method change you suggest in #1 is still a breaking change. I totally agree people should be ok with breaking changes but honestly, I rather first make the change in a non-breaking way and then we can always later make the breaking change to clean up the api

Reply:
Ok, I'll try to do some kind of workaround to get rid of breaking changes. 

By the way, why replacing those interface's methods with extension methods for the interface is a breaking change? If we keep it in the same namespace, everything should compile without the need for any code adjustments?

Reply:
I think I've made all the necessary changes to avoid breaking changes. I deleted some tests because they do not any value from the "business" perspective (they were testing implementation details and there was no sense to fix them after changing those details).

Can you take a look a the new changes? I Rebase everything to discard previous changes.